File: test_scripts/python/run_client.py
import sys
import getopt
import json

def getErrorMessage(error):
    error_type = type(error).__name__
    message = error.__str__()
    if error_type == 'ServerError':
        message = error.data
    return message

def main(argv):
    sys.path.append('./')
    endpoint = None
    token = None
    async_job_check_time_ms = None
    module_file = None
    class_name = None
    method_name = None
    input_filepath = None
    output_filepath = None
    error_filepath = None
    try:
        opts, args = getopt.getopt(argv,"he:t:a:g:c:m:i:o:r:",["help","endpoint=","token=","asyncchecktime=",
                                                                "package=","class=","method=","input=","output=","error="])
    except getopt.GetoptError:
        print('Please use "test_client.py -h" or "test_client.py --help" for help')
        sys.exit(2)
    
    for opt, arg in opts:
        if opt in ("-h", "--help"):
            print('run_client.py --input <json_file> --endpoint <url> [--token <kbase_token> [--asyncchecktime <ms>]]')
            sys.exit()
        elif opt in ("-e", "--endpoint"):
            endpoint = arg
        elif opt in ("-t", "--token"):
            token = arg
        elif opt in ("-a", "--asyncchecktime"):
            try:
                async_job_check_time_ms = long(arg)
            except NameError:
                async_job_check_time_ms = int(arg)
        elif opt in ("-g", "--package"):
            module_file = arg
        elif opt in ("-c", "--class"):
            class_name = arg
        elif opt in ("-m", "--method"):
            method_name = arg
        elif opt in ("-i", "--input"):
            input_filepath = arg
        elif opt in ("-o", "--output"):
            output_filepath = arg
        elif opt in ("-r", "--error"):
            error_filepath = arg
    inp_file = open(input_filepath, "r")
    params = json.load(inp_file)
    inp_file.close()
    module = __import__(module_file)
    client_class = getattr(module, class_name)
    client_instance = None
    if user is not None:
        if async_job_check_time_ms:
            client_instance = client_class(url = endpoint, token = token, async_job_check_time_ms = async_job_check_time_ms)
        else:
            client_instance = client_class(url = endpoint, token = token)
    else:
        client_instance = client_class(url = endpoint, ignore_authrc = True)
    method_instance = getattr(client_instance, method_name)
    try:
        ret = method_instance(*params, context={})
        with open(output_filepath, "w") as out_file:
            out_file.write(json.dumps(ret, sort_keys=True))
    except Exception as ex:
        with open(error_filepath, "w") as err_file:
            err_file.write(getErrorMessage(ex))
        

if __name__ == "__main__":
    main(sys.argv[1:])

File: test_scripts/perl/run-client.pl
#
# Automatic test rig for KBase perl client generation.
#
#
#
#
use strict;
use warnings;

use JSON;
use Data::Dumper;
use Getopt::Long;

my $DESCRIPTION =
"
USAGE
      perl test-client.pl [options]
      
DESCRIPTION
      Conducts an automatic test of a perl client against the specified endpoint using
      methods and parameters indicated in the tests config file.
      
      --endpoint [url]      endpoint of the server to test
      --token [token]       token for testing authenticated calls
      --asyncchecktime [ms] time client waits every cycle of checking state of async methods 
      --package [package]   package with client module
      --method [method]     method to run
      --input [filename]    input file with parameters in JSON
      --output [filename]   output file with results in JSON
      --error [filename]    output file with text error message
      -h, --help            display this help message, ignore all arguments
";
      
# first parse options to get the testconfig file
my $verbose;
my $endpoint;

my $token;
my $async_job_check_time_ms;
my $client_module;
my $method;
my $input_filename;
my $output_filename;
my $error_filename;

my $help;

my $opt = GetOptions (
        "verbose|v" => \$verbose,
        "endpoint=s" => \$endpoint,
        "token=s" => \$token,
        "asyncchecktime=i" => \$async_job_check_time_ms,
        "help|h" => \$help,
        "package=s" => \$client_module,
        "method=s" => \$method,
        "input=s" => \$input_filename,
        "output=s" => \$output_filename,
        "error=s" => \$error_filename,
        );

if ($help) {
    print $DESCRIPTION."\n";
    exit 0;
}


# endpoint must be defined
if (!$endpoint) {
    fail("endpoint parameter must be defined");
    exit 1;
}

# package must be defined
if (!$client_module) {
    fail("package parameter must be defined");
    exit 1;
}

# method must be defined
if (!$method) {
    fail("method parameter must be defined");
    exit 1;
}

# input must be defined
if (!$input_filename) {
    fail("input parameter must be defined");
    exit 1;
}

# output must be defined
if (!$output_filename) {
    fail("output parameter must be defined");
    exit 1;
}

# error must be defined
if (!$error_filename) {
    fail("error parameter must be defined");
    exit 1;
}

#parse the tests
open(my $fh, "<", $input_filename);
my $input_string='';
while (my $line = <$fh>) {
    chomp $line;
    $input_string .= $line;
}
close($fh);
my $params = JSON->new->decode($input_string);

my $json = JSON->new->canonical;

eval("use $client_module;");
#instantiate an authenticated client and a nonauthenticated client
my $client;

if($token) {
    if ($async_job_check_time_ms) {
        $client=$client_module->new($endpoint, token=>$token, async_job_check_time_ms=>$async_job_check_time_ms);
    } else {
        $client=$client_module->new($endpoint, token=>$token);
    }
} else {
    $client = $client_module->new($endpoint,ignore_kbase_config=>1);
}

# loop over each test case and run it against the server.  We create a new client instance
# for each test
    
my @result;
no strict "refs";
eval {
    @result = $client->$method(@{$params});
};
if($@) {
    print "caught: $@";
    my $returnedErrorMsg = "$@";
    my $OUTFILE;
    open $OUTFILE, '>>', $error_filename;
    print { $OUTFILE } $returnedErrorMsg;
    close $OUTFILE;
} else {
    my $serialized_result = $json->encode(\@result);
    my $OUTFILE;
    open $OUTFILE, '>>', $output_filename;
    print { $OUTFILE } $serialized_result;
    close $OUTFILE;
}





